torchgeo/datasets/splits.py

```diff
@@ -0,0 +1,359 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""Dataset splitting utilities."""
+
+from copy import deepcopy
+from itertools import accumulate
+from math import floor
+from typing import List, Optional, Sequence, Tuple, Union
+
+from rtree.index import Index, Property
+from torch import Generator, default_generator, randint, randperm
+
+from ..datasets import GeoDataset
+from .utils import BoundingBox
+
+__all__ = (
+    "random_bbox_assignment",
+    "random_bbox_splitting",
+    "random_grid_cell_assignment",
+    "roi_split",
+    "time_series_split",
+)
+
+
+def _fractions_to_lengths(fractions: Sequence[float], total: int) -> Sequence[int]:
+    """Utility to divide a number into a list of integers according to fractions.
+
+    Args:
+        fractions: list of fractions
+        total: total to be divided
+
+    Returns:
+        List of lengths.
+
+    .. versionadded:: 0.5
+    """
+    lengths = [floor(frac * total) for frac in fractions]
+    remainder = int(total - sum(lengths))
+    # add 1 to all the lengths in round-robin fashion until the remainder is 0
+    for i in range(remainder):
+        idx_to_add_at = i % len(lengths)
+        lengths[idx_to_add_at] += 1
+    return lengths
+
+
+def random_bbox_assignment(
+    dataset: GeoDataset,
+    lengths: Sequence[Union[int, float]],
+    generator: Optional[Generator] = default_generator,
+) -> List[GeoDataset]:
+    """Split a GeoDataset randomly assigning its index's BoundingBoxes.
+
+    Args:
+        dataset: dataset to be split
+        lengths: lengths or fractions of splits to be produced
+        generator: (optional) generator used for the random permutation
+
+    Returns
+        A list of the subset datasets.
+
+    .. versionadded:: 0.5
+    """
+    if not (sum(lengths) == 1 or sum(lengths) == len(dataset)):
+        raise ValueError(
+            "Sum of input lengths must equal 1 or the length of dataset's index."
+        )
+
+    if any(n <= 0 for n in lengths):
+        raise ValueError("All items in input lengths must be greater than 0.")
+
+    if sum(lengths) == 1:
+        lengths = _fractions_to_lengths(lengths, len(dataset))
```

Comment:
I'm surprised typing doesn't complain about this since you're converting `Sequence[float]` to `Sequence[int]`.

Reply:
Oh wait, that's why you have `ignore[arg-type]` everywhere. Can we create a new variable or explicitly cast to `Sequence[int]`?

Reply:
Yeah all the `ignore[arg-type]` are because we are supporting different types in lengths, which are used in different ways. I'm not sure how to handle it properly for mypy, specially in time_series_split were there could be numbers or tuples.

I prefer having only one input though.

Reply:
I managed to get rid of most ignores using mypy's cast function. Only one remained in line 316 as mypy is still complaining about types there for some reason.